File: scripts/translate-srd.mjs
/**
 * translate-srd.mjs
 * 
 * Script para traducir al español todos los registros del SRD en la base de datos.
 * Usa la API de Gemini Flash para traducir en batch, con progreso y reanudación.
 * 
 * Uso:
 *   node scripts/translate-srd.mjs              → Traduce todo
 *   node scripts/translate-srd.mjs --type monster  → Solo monstruos
 *   node scripts/translate-srd.mjs --type spell    → Solo hechizos
 *   node scripts/translate-srd.mjs --type item     → Solo objetos
 *   node scripts/translate-srd.mjs --reset         → Marca todo como no traducido (reinicia)
 */

import Database from 'better-sqlite3';
import { GoogleGenAI } from '@google/genai';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const dbPath = path.join(__dirname, '../vtt_database.db');

// ---- CONFIG ----
const GEMINI_API_KEY = process.env.GEMINI_API_KEY || 'AIzaSyA-nKRgmWgDr2XXzWPCUwUWaj6QHTf4uKo';
const BATCH_SIZE = 5;          // Cuántos registros por llamada a Gemini
const DELAY_MS = 1500;         // Pausa entre llamadas para respetar rate limits
const MODEL = 'gemini-2.0-flash'; // Modelo rápido y económico

const db = new Database(dbPath);
const genAI = new GoogleGenAI({ apiKey: GEMINI_API_KEY });

// Leer argumentos
const args = process.argv.slice(2);
const typeFilter = args.includes('--type') ? args[args.indexOf('--type') + 1] : null;
const doReset = args.includes('--reset');

// ---- SETUP: añadir columna 'translated' si no existe ----
try {
  db.exec("ALTER TABLE content_items ADD COLUMN translated INTEGER DEFAULT 0");
  console.log("✅ Columna 'translated' añadida a la base de datos.");
} catch (e) {
  // Ya existe, ignorar
}

if (doReset) {
  const q = typeFilter
    ? db.prepare("UPDATE content_items SET translated = 0 WHERE type = ?").run(typeFilter)
    : db.prepare("UPDATE content_items SET translated = 0").run();
  console.log(`🔄 Reinicio: ${q.changes} registros marcados como no traducidos.`);
  process.exit(0);
}

// ---- SISTEMA DE PROGRESO ----
function getProgress() {
  const total = typeFilter
    ? db.prepare("SELECT COUNT(*) as c FROM content_items WHERE type = ?").get(typeFilter)
    : db.prepare("SELECT COUNT(*) as c FROM content_items").get();
  const done = typeFilter
    ? db.prepare("SELECT COUNT(*) as c FROM content_items WHERE translated = 1 AND type = ?").get(typeFilter)
    : db.prepare("SELECT COUNT(*) as c FROM content_items WHERE translated = 1").get();
  return { total: total.c, done: done.c };
}

function getNextBatch() {
  const query = typeFilter
    ? "SELECT id, name, type, data FROM content_items WHERE translated = 0 AND type = ? LIMIT ?"
    : "SELECT id, name, type, data FROM content_items WHERE translated = 0 LIMIT ?";
  return typeFilter
    ? db.prepare(query).all(typeFilter, BATCH_SIZE)
    : db.prepare(query).all(BATCH_SIZE);
}

function markTranslated(id) {
  db.prepare("UPDATE content_items SET translated = 1 WHERE id = ?").run(id);
}

function updateRecord(id, name, data) {
  db.prepare("UPDATE content_items SET name = ?, data = ?, translated = 1 WHERE id = ?")
    .run(name, JSON.stringify(data), id);
}

// ---- LÓGICA DE TRADUCCIÓN POR TIPO ----

function extractTextsToTranslate(type, name, data) {
  /**
   * Devuelve un objeto con los campos a traducir.
   * Los campos numéricos, slugs y URLs se omiten.
   */
  const texts = { name };

  if (type === 'monster') {
    if (data.description) texts.description = data.description;
    if (data.size) texts.size = data.size;
    if (data.type) texts.creature_type = data.type; // "type" es la raza del monstruo (Humanoid, Beast, etc.)
    if (Array.isArray(data.actions) && data.actions.length > 0) {
      texts.actions = data.actions.map(a => ({
        name: a.name || '',
        desc: a.desc || a.description || ''
      }));
    }
    if (Array.isArray(data.traits) && data.traits.length > 0) {
      texts.traits = data.traits.map(t => ({
        name: t.name || '',
        desc: t.desc || t.description || ''
      }));
    }
    if (Array.isArray(data.vulnerabilities) && data.vulnerabilities.length > 0) texts.vulnerabilities = data.vulnerabilities;
    if (Array.isArray(data.resistances) && data.resistances.length > 0) texts.resistances = data.resistances;
    if (Array.isArray(data.immunities) && data.immunities.length > 0) texts.immunities = data.immunities;
    if (Array.isArray(data.legendary_actions) && data.legendary_actions.length > 0) {
      texts.legendary_actions = data.legendary_actions.map(a => ({
        name: a.name || '',
        desc: a.desc || a.description || ''
      }));
    }
    if (Array.isArray(data.special_abilities) && data.special_abilities.length > 0) {
      texts.special_abilities = data.special_abilities.map(a => ({
        name: a.name || '',
        desc: a.desc || a.description || ''
      }));
    }
    if (Array.isArray(data.reactions) && data.reactions.length > 0) {
      texts.reactions = data.reactions.map(a => ({
        name: a.name || '',
        desc: a.desc || a.description || ''
      }));
    }

  } else if (type === 'spell') {
    if (data.desc) texts.desc = Array.isArray(data.desc) ? data.desc.join('\n') : data.desc;
    if (data.higher_level) texts.higher_level = Array.isArray(data.higher_level) ? data.higher_level.join('\n') : data.higher_level;
    if (data.components) texts.components = data.components;
    if (data.material) texts.material = data.material;
    if (data.duration) texts.duration = data.duration;
    if (data.casting_time) texts.casting_time = data.casting_time;
    if (data.school) texts.school = typeof data.school === 'object' ? data.school.name : data.school;
    if (data.range) texts.range = data.range;

  } else if (type === 'item') {
    if (data.desc) texts.desc = Array.isArray(data.desc) ? data.desc.join('\n') : data.desc;
    if (data.description) texts.description = data.description;
    if (data.equipment_category) texts.equipment_category = typeof data.equipment_category === 'object' ? data.equipment_category.name : data.equipment_category;
    if (data.weapon_category) texts.weapon_category = data.weapon_category;
    if (data.weapon_range) texts.weapon_range = data.weapon_range;
    if (data.armor_category) texts.armor_category = data.armor_category;
    if (data.rarity) texts.rarity = typeof data.rarity === 'object' ? data.rarity.name : data.rarity;
    if (data.tool_category) texts.tool_category = data.tool_category;
  }

  return texts;
}

function applyTranslations(type, originalData, translations, translatedName) {
  /**
   * Aplica las traducciones al objeto de datos original,
   * manteniendo intactos los campos numéricos/técnicos.
   */
  const d = { ...originalData };

  if (type === 'monster') {
    if (translations.description) d.description = translations.description;
    if (translations.size) d.size = translations.size;
    if (translations.creature_type) d.type = translations.creature_type;
    if (translations.actions) {
      d.actions = (originalData.actions || []).map((a, i) => ({
        ...a,
        name: translations.actions[i]?.name || a.name,
        desc: translations.actions[i]?.desc || a.desc || a.description,
        description: translations.actions[i]?.desc || a.desc || a.description,
      }));
    }
    if (translations.traits) {
      d.traits = (originalData.traits || []).map((t, i) => ({
        ...t,
        name: translations.traits[i]?.name || t.name,
        desc: translations.traits[i]?.desc || t.desc || t.description,
      }));
    }
    if (translations.vulnerabilities) d.vulnerabilities = translations.vulnerabilities;
    if (translations.resistances) d.resistances = translations.resistances;
    if (translations.immunities) d.immunities = translations.immunities;
    if (translations.legendary_actions) {
      d.legendary_actions = (originalData.legendary_actions || []).map((a, i) => ({
        ...a,
        name: translations.legendary_actions[i]?.name || a.name,
        desc: translations.legendary_actions[i]?.desc || a.desc || a.description,
      }));
    }
    if (translations.special_abilities) {
      d.special_abilities = (originalData.special_abilities || []).map((a, i) => ({
        ...a,
        name: translations.special_abilities[i]?.name || a.name,
        desc: translations.special_abilities[i]?.desc || a.desc || a.description,
      }));
    }
    if (translations.reactions) {
      d.reactions = (originalData.reactions || []).map((a, i) => ({
        ...a,
        name: translations.reactions[i]?.name || a.name,
        desc: translations.reactions[i]?.desc || a.desc || a.description,
      }));
    }

  } else if (type === 'spell') {
    if (translations.desc) d.desc = [translations.desc];
    if (translations.higher_level) d.higher_level = [translations.higher_level];
    if (translations.duration) d.duration = translations.duration;
    if (translations.casting_time) d.casting_time = translations.casting_time;
    if (translations.school) {
      if (typeof originalData.school === 'object') {
        d.school = { ...originalData.school, name: translations.school };
      } else {
        d.school = translations.school;
      }
    }
    if (translations.range) d.range = translations.range;
    if (translations.material) d.material = translations.material;
    if (translations.components) d.components = translations.components;

  } else if (type === 'item') {
    if (translations.desc) d.desc = [translations.desc];
    if (translations.description) d.description = translations.description;
    if (translations.equipment_category) {
      if (typeof originalData.equipment_category === 'object') {
        d.equipment_category = { ...originalData.equipment_category, name: translations.equipment_category };
      } else {
        d.equipment_category = translations.equipment_category;
      }
    }
    if (translations.weapon_category) d.weapon_category = translations.weapon_category;
    if (translations.weapon_range) d.weapon_range = translations.weapon_range;
    if (translations.armor_category) d.armor_category = translations.armor_category;
    if (translations.rarity) {
      if (typeof originalData.rarity === 'object') {
        d.rarity = { ...originalData.rarity, name: translations.rarity };
      } else {
        d.rarity = translations.rarity;
      }
    }
    if (translations.tool_category) d.tool_category = translations.tool_category;
  }

  return d;
}

// ---- LLAMADA A GEMINI ----
async function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

async function translateBatch(items, retryCount = 0) {
  const MAX_RETRIES = 5;
  const prompt = `Eres un experto traductor de contenido de Dungeons & Dragons 5e del inglés al español latinoamericano. 
Usa la terminología oficial de D&D en español (ej: "Sneak Attack" → "Ataque Furtivo", "Saving Throw" → "Tirada de Salvación", "Hit Points" → "Puntos de Golpe", "Armor Class" → "Clase de Armadura", etc.).
Mantén los nombres propios que no tienen traducción estándar en inglés.
Devuelve ÚNICAMENTE un JSON válido, sin explicaciones ni markdown, con este formato exacto:
[
  { "id": <id>, "name": "<nombre traducido>", "translations": { <campos traducidos> } },
  ...
]

Ítems a traducir:
${JSON.stringify(items.map(item => ({ id: item.id, type: item.type, texts: item.texts })), null, 2)}

IMPORTANTE:
- Para campos que son arrays (como actions, traits), devuelve el array con los mismos índices.
- Para vulnerabilities/resistances/immunities que son arrays de strings, devuelve arrays de strings traducidos.
- No omitas ningún campo que se te dio para traducir.
- Responde SOLO con el JSON array, sin texto adicional.`;

  try {
    const result = await genAI.models.generateContent({
      model: MODEL,
      contents: [{ role: 'user', parts: [{ text: prompt }] }]
    });
    const text = result.candidates?.[0]?.content?.parts?.[0]?.text?.trim() || '';
    
    // Limpiar posible markdown
    const cleaned = text
      .replace(/^```json\s*/i, '')
      .replace(/^```\s*/i, '')
      .replace(/\s*```$/i, '')
      .trim();
    
    return JSON.parse(cleaned);
  } catch (e) {
    const msg = e.message || '';
    const is429 = msg.includes('429') || msg.includes('quota') || msg.includes('rate');
    
    if (is429 && retryCount < MAX_RETRIES) {
      const waitSec = Math.pow(2, retryCount + 2); // 4s, 8s, 16s, 32s, 64s
      process.stdout.write(`\n    ⏳ Rate limit (429), esperando ${waitSec}s antes de reintentar (intento ${retryCount + 1}/${MAX_RETRIES})... `);
      await sleep(waitSec * 1000);
      return translateBatch(items, retryCount + 1);
    }
    
    console.error('  ❌ Error en Gemini:', msg.substring(0, 200));
    return null;
  }
}

// ---- MAIN ----
async function main() {
  console.log('\n🌐 TRADUCTOR SRD → ESPAÑOL (powered by Gemini)');
  console.log('══════════════════════════════════════════════');
  
  const { total, done } = getProgress();
  console.log(`📊 Registros totales: ${total}`);
  console.log(`✅ Ya traducidos:     ${done}`);
  console.log(`⏳ Pendientes:        ${total - done}`);
  if (typeFilter) console.log(`🔍 Filtrando tipo:    ${typeFilter}`);
  console.log('');

  if (total - done === 0) {
    console.log('🎉 ¡Todo ya está traducido!');
    return;
  }

  let processed = 0;
  let errors = 0;
  const startTime = Date.now();

  while (true) {
    const batch = getNextBatch();
    if (batch.length === 0) break;

    // Preparar datos para traducción
    const itemsToTranslate = batch.map(row => {
      let data;
      try {
        data = JSON.parse(row.data);
      } catch {
        data = {};
      }
      return {
        id: row.id,
        type: row.type,
        name: row.name,
        data,
        texts: extractTextsToTranslate(row.type, row.name, data)
      };
    });

    // Llamar a Gemini
    process.stdout.write(`  📤 Traduciendo batch [${done + processed + 1}-${done + processed + batch.length}/${total}]... `);
    const translations = await translateBatch(itemsToTranslate);

    if (!translations) {
      // En caso de error, marcar como traducido para no bloquearse
      batch.forEach(row => markTranslated(row.id));
      errors += batch.length;
      console.log(`⚠️ Error en batch, saltando.`);
    } else {
      // Aplicar traducciones
      for (const t of translations) {
        const original = itemsToTranslate.find(i => i.id === t.id);
        if (!original) continue;
        
        const translatedData = applyTranslations(
          original.type,
          original.data,
          t.translations || {},
          t.name
        );
        updateRecord(t.id, t.name, translatedData);
      }
      
      // Marcar los que Gemini no devolvió (por si faltó alguno)
      batch.forEach(row => {
        const found = translations.find(t => t.id === row.id);
        if (!found) markTranslated(row.id);
      });

      processed += batch.length;
      const elapsed = ((Date.now() - startTime) / 1000).toFixed(0);
      const rate = (processed / elapsed * 60).toFixed(0);
      console.log(`✓ (${rate} reg/min, ${elapsed}s transcurridos)`);
    }

    // Pausa entre requests para no exceder rate limits
    await new Promise(r => setTimeout(r, DELAY_MS));
  }

  const totalTime = ((Date.now() - startTime) / 1000 / 60).toFixed(1);
  console.log('\n══════════════════════════════════════════════');
  console.log(`🎉 ¡Traducción completada!`);
  console.log(`   ✅ Traducidos: ${processed}`);
  console.log(`   ⚠️ Errores:    ${errors}`);
  console.log(`   ⏱️ Tiempo:     ${totalTime} minutos`);
  console.log('══════════════════════════════════════════════\n');
}

main().catch(console.error);

File: scripts/translate_srd.py
# -*- coding: utf-8 -*-
import sqlite3
import json
import os
import urllib.request
import urllib.error
import time

db_path = r"c:\Users\Sapo\Documents\DND\dnd-vtt\vtt_database.db"
env_path = r"c:\Users\Sapo\Documents\DND\dnd-vtt\.env"

# Load API key
api_key = "AIzaSyA-nKRgmWgDr2XXzWPCUwUWaj6QHTf4uKo"
if os.path.exists(env_path):
    with open(env_path, 'r', encoding='utf-8') as f:
        for line in f:
            if line.startswith("GEMINI_API_KEY="):
                api_key = line.split("=")[1].strip()

MODEL = "gemini-2.0-flash"
BATCH_SIZE = 5
DELAY_SEC = 6.0

print("--- Iniciando Script de Traduccion D&D 5e (Python + Gemini) ---")

conn = sqlite3.connect(db_path)
cursor = conn.cursor()

# Ensure 'translated' column exists
cursor.execute("PRAGMA table_info(content_items)")
columns = [col[1] for col in cursor.fetchall()]
if 'translated' not in columns:
    cursor.execute("ALTER TABLE content_items ADD COLUMN translated INTEGER DEFAULT 0")
    conn.commit()
    print("Columna 'translated' agregada a content_items.")

def get_progress():
    cursor.execute("SELECT COUNT(*) FROM content_items WHERE type IN ('monster', 'spell', 'item')")
    total = cursor.fetchone()[0]
    cursor.execute("SELECT COUNT(*) FROM content_items WHERE type IN ('monster', 'spell', 'item') AND translated = 1")
    done = cursor.fetchone()[0]
    return total, done

def get_next_batch():
    cursor.execute(
        "SELECT id, name, type, data FROM content_items WHERE type IN ('monster', 'spell', 'item') AND translated = 0 LIMIT ?",
        (BATCH_SIZE,)
    )
    return cursor.fetchall()

def update_record(record_id, name, data_str):
    cursor.execute(
        "UPDATE content_items SET name = ?, data = ?, translated = 1 WHERE id = ?",
        (name, data_str, record_id)
    )

def mark_translated(record_id):
    cursor.execute("UPDATE content_items SET translated = 1 WHERE id = ?", (record_id,))

def extract_texts_to_translate(item_type, name, data):
    texts = {"name": name}
    if item_type == 'monster':
        if 'description' in data: texts['description'] = data['description']
        if 'size' in data: texts['size'] = data['size']
        if 'type' in data: texts['creature_type'] = data['type']
        if 'actions' in data and isinstance(data['actions'], list):
            texts['actions'] = [{"name": a.get('name', ''), "desc": a.get('desc') or a.get('description', '')} for a in data['actions']]
        if 'traits' in data and isinstance(data['traits'], list):
            texts['traits'] = [{"name": t.get('name', ''), "desc": t.get('desc') or t.get('description', '')} for t in data['traits']]
        if 'vulnerabilities' in data: texts['vulnerabilities'] = data['vulnerabilities']
        if 'resistances' in data: texts['resistances'] = data['resistances']
        if 'immunities' in data: texts['immunities'] = data['immunities']
        if 'legendary_actions' in data and isinstance(data['legendary_actions'], list):
            texts['legendary_actions'] = [{"name": a.get('name', ''), "desc": a.get('desc') or a.get('description', '')} for a in data['legendary_actions']]
        if 'special_abilities' in data and isinstance(data['special_abilities'], list):
            texts['special_abilities'] = [{"name": a.get('name', ''), "desc": a.get('desc') or a.get('description', '')} for a in data['special_abilities']]
        if 'reactions' in data and isinstance(data['reactions'], list):
            texts['reactions'] = [{"name": a.get('name', ''), "desc": a.get('desc') or a.get('description', '')} for a in data['reactions']]
    elif item_type == 'spell':
        if 'desc' in data:
            texts['desc'] = "\n".join(data['desc']) if isinstance(data['desc'], list) else data['desc']
        if 'higher_level' in data:
            texts['higher_level'] = "\n".join(data['higher_level']) if isinstance(data['higher_level'], list) else data['higher_level']
        if 'components' in data: texts['components'] = data['components']
        if 'material' in data: texts['material'] = data['material']
        if 'duration' in data: texts['duration'] = data['duration']
        if 'casting_time' in data: texts['casting_time'] = data['casting_time']
        if 'school' in data:
            texts['school'] = data['school'].get('name') if isinstance(data['school'], dict) else data['school']
        if 'range' in data: texts['range'] = data['range']
    elif item_type == 'item':
        if 'desc' in data:
            texts['desc'] = "\n".join(data['desc']) if isinstance(data['desc'], list) else data['desc']
        if 'description' in data: texts['description'] = data['description']
        if 'equipment_category' in data:
            texts['equipment_category'] = data['equipment_category'].get('name') if isinstance(data['equipment_category'], dict) else data['equipment_category']
        if 'weapon_category' in data: texts['weapon_category'] = data['weapon_category']
        if 'weapon_range' in data: texts['weapon_range'] = data['weapon_range']
        if 'armor_category' in data: texts['armor_category'] = data['armor_category']
        if 'rarity' in data:
            texts['rarity'] = data['rarity'].get('name') if isinstance(data['rarity'], dict) else data['rarity']
        if 'tool_category' in data: texts['tool_category'] = data['tool_category']
    return texts

def apply_translations(item_type, original_data, translations):
    d = original_data.copy()
    if item_type == 'monster':
        if 'description' in translations: d['description'] = translations['description']
        if 'size' in translations: d['size'] = translations['size']
        if 'creature_type' in translations: d['type'] = translations['creature_type']
        if 'actions' in translations:
            d['actions'] = []
            orig_actions = original_data.get('actions') or []
            for i, a in enumerate(translations['actions']):
                orig = orig_actions[i] if i < len(orig_actions) else {}
                new_a = orig.copy()
                new_a['name'] = a.get('name') or orig.get('name', '')
                new_a['desc'] = a.get('desc') or orig.get('desc') or orig.get('description', '')
                new_a['description'] = new_a['desc']
                d['actions'].append(new_a)
        if 'traits' in translations:
            d['traits'] = []
            orig_traits = original_data.get('traits') or []
            for i, t in enumerate(translations['traits']):
                orig = orig_traits[i] if i < len(orig_traits) else {}
                new_t = orig.copy()
                new_t['name'] = t.get('name') or orig.get('name', '')
                new_t['desc'] = t.get('desc') or orig.get('desc') or orig.get('description', '')
                d['traits'].append(new_t)
        if 'vulnerabilities' in translations: d['vulnerabilities'] = translations['vulnerabilities']
        if 'resistances' in translations: d['resistances'] = translations['resistances']
        if 'immunities' in translations: d['immunities'] = translations['immunities']
        if 'legendary_actions' in translations:
            d['legendary_actions'] = []
            orig_legendary = original_data.get('legendary_actions') or []
            for i, a in enumerate(translations['legendary_actions']):
                orig = orig_legendary[i] if i < len(orig_legendary) else {}
                new_a = orig.copy()
                new_a['name'] = a.get('name') or orig.get('name', '')
                new_a['desc'] = a.get('desc') or orig.get('desc') or orig.get('description', '')
                d['legendary_actions'].append(new_a)
        if 'special_abilities' in translations:
            d['special_abilities'] = []
            orig_special = original_data.get('special_abilities') or []
            for i, a in enumerate(translations['special_abilities']):
                orig = orig_special[i] if i < len(orig_special) else {}
                new_a = orig.copy()
                new_a['name'] = a.get('name') or orig.get('name', '')
                new_a['desc'] = a.get('desc') or orig.get('desc') or orig.get('description', '')
                d['special_abilities'].append(new_a)
        if 'reactions' in translations:
            d['reactions'] = []
            orig_reactions = original_data.get('reactions') or []
            for i, a in enumerate(translations['reactions']):
                orig = orig_reactions[i] if i < len(orig_reactions) else {}
                new_a = orig.copy()
                new_a['name'] = a.get('name') or orig.get('name', '')
                new_a['desc'] = a.get('desc') or orig.get('desc') or orig.get('description', '')
                d['reactions'].append(new_a)
    elif item_type == 'spell':
        if 'desc' in translations: d['desc'] = [translations['desc']]
        if 'higher_level' in translations: d['higher_level'] = [translations['higher_level']]
        if 'duration' in translations: d['duration'] = translations['duration']
        if 'casting_time' in translations: d['casting_time'] = translations['casting_time']
        if 'school' in translations:
            if isinstance(original_data.get('school'), dict):
                d['school'] = original_data['school'].copy()
                d['school']['name'] = translations['school']
            else:
                d['school'] = translations['school']
        if 'range' in translations: d['range'] = translations['range']
        if 'material' in translations: d['material'] = translations['material']
        if 'components' in translations: d['components'] = translations['components']
    elif item_type == 'item':
        if 'desc' in translations: d['desc'] = [translations['desc']]
        if 'description' in translations: d['description'] = translations['description']
        if 'equipment_category' in translations:
            if isinstance(original_data.get('equipment_category'), dict):
                d['equipment_category'] = original_data['equipment_category'].copy()
                d['equipment_category']['name'] = translations['equipment_category']
            else:
                d['equipment_category'] = translations['equipment_category']
        if 'weapon_category' in translations: d['weapon_category'] = translations['weapon_category']
        if 'weapon_range' in translations: d['weapon_range'] = translations['weapon_range']
        if 'armor_category' in translations: d['armor_category'] = translations['armor_category']
        if 'rarity' in translations:
            if isinstance(original_data.get('rarity'), dict):
                d['rarity'] = original_data['rarity'].copy()
                d['rarity']['name'] = translations['rarity']
            else:
                d['rarity'] = translations['rarity']
        if 'tool_category' in translations: d['tool_category'] = translations['tool_category']
    return d

def translate_batch(items, retry_count=0):
    MAX_RETRIES = 5
    prompt = (
        "Eres un experto traductor de contenido de Dungeons & Dragons 5e del ingles al espanol latinoamericano.\n"
        "Usa la terminologia oficial de D&D en espanol (ej: 'Sneak Attack' -> 'Ataque Furtivo', 'Saving Throw' -> 'Tirada de Salvacion', 'Hit Points' -> 'Puntos de Golpe', 'Armor Class' -> 'Clase de Armadura', etc.).\n"
        "Manten los nombres propios que no tienen traduccion estandar en ingles.\n"
        "Devuelve UNICAMENTE un JSON valido, sin explicaciones ni markdown, con este formato exacto:\n"
        "[\n"
        "  { \"id\": <id>, \"name\": \"<nombre traducido>\", \"translations\": { <campos traducidos> } },\n"
        "  ...\n"
        "]\n\n"
        "Items a traducir:\n" +
        json.dumps([{"id": item["id"], "type": item["type"], "texts": item["texts"]} for item in items], indent=2) +
        "\n\nIMPORTANTE:\n"
        "- Para campos que son arrays (como actions, traits), devuelve el array con los mismos indices.\n"
        "- Para vulnerabilities/resistances/immunities que son arrays de strings, devuelve arrays de strings traducidos.\n"
        "- No omitas ningun campo que se te dio para traducir.\n"
        "- Responde SOLO con el JSON array, sin texto adicional ni markdown."
    )

    url = f"https://generativelanguage.googleapis.com/v1beta/models/{MODEL}:generateContent?key={api_key}"
    req_body = {
        "contents": [
            {
                "parts": [
                    {"text": prompt}
                ]
            }
        ]
    }
    
    headers = {"Content-Type": "application/json"}
    
    try:
        req = urllib.request.Request(url, data=json.dumps(req_body).encode('utf-8'), headers=headers, method='POST')
        with urllib.request.urlopen(req, timeout=45) as response:
            res_data = json.loads(response.read().decode('utf-8'))
            
        candidates = res_data.get('candidates', [])
        if not candidates:
            return None
            
        text = candidates[0].get('content', {}).get('parts', [])[0].get('text', '').strip()
        cleaned = text.replace("```json", "").replace("```", "").strip()
        return json.loads(cleaned)
    except Exception as e:
        err_msg = str(e)
        is_429 = "429" in err_msg or "quota" in err_msg or "rate" in err_msg
        
        if is_429 and retry_count < MAX_RETRIES:
            wait_sec = 2 ** (retry_count + 2)
            print(f"\n   [Rate limit 429] Esperando {wait_sec}s antes de reintentar (intento {retry_count + 1}/{MAX_RETRIES})...")
            time.sleep(wait_sec)
            return translate_batch(items, retry_count + 1)
            
        print(f"\n   Error en Gemini API: {err_msg[:120]}")
        return None

total, done = get_progress()
print(f"Registros totales: {total}")
print(f"Ya traducidos:     {done}")
print(f"Pendientes:        {total - done}")

if total - done <= 0:
    print("Todo ya esta traducido. Finalizando.")
    conn.close()
    exit(0)

processed = 0
errors = 0
start_time = time.time()

try:
    while True:
        batch = get_next_batch()
        if not batch:
            break

        items_to_translate = []
        for row_id, name, item_type, data_str in batch:
            try:
                data = json.loads(data_str)
            except Exception:
                data = {}
            
            items_to_translate.append({
                "id": row_id,
                "type": item_type,
                "name": name,
                "data": data,
                "texts": extract_texts_to_translate(item_type, name, data)
            })

        print(f"  Traduciendo batch [{done + processed + 1}-{min(total, done + processed + len(batch))}/{total}]...", end="", flush=True)
        translations = translate_batch(items_to_translate)

        if not translations:
            print("\n\n[ERROR DE CUOTA / CONEXIÓN]")
            print("No se pudo obtener respuesta de la API de Gemini tras múltiples intentos de reintento.")
            print("Esto indica que has agotado la cuota de uso (por minuto o diaria) de tu clave API gratuita de Gemini,")
            print("o bien que hay un problema temporal de conexión con el servicio de Google.")
            print("El script se detendrá de forma segura AHORA SIN marcar estos registros como traducidos,")
            print("de modo que puedas continuar exactamente desde aquí más tarde al restablecerse tu cuota.")
            print("Intenta ejecutar de nuevo el script en unos minutos o en unas horas.")
            conn.close()
            exit(1)
        else:
            for t in translations:
                t_id = t.get('id')
                t_name = t.get('name')
                t_fields = t.get('translations') or {}
                
                original = next((i for i in items_to_translate if i["id"] == t_id), None)
                if not original:
                    continue
                
                translated_data = apply_translations(original["type"], original["data"], t_fields)
                update_record(t_id, t_name, json.dumps(translated_data, ensure_ascii=False))

            # Mark missing ones in translation
            for row_id, _, _, _ in batch:
                found = next((t for t in translations if t.get('id') == row_id), None)
                if not found:
                    mark_translated(row_id)

            processed += len(batch)
            elapsed = int(time.time() - start_time)
            rate = int(processed / elapsed * 60) if elapsed > 0 else 0
            print(f" [OK] ({rate} reg/min, {elapsed}s transcurridos)")
            conn.commit()

        time.sleep(DELAY_SEC)

    elapsed_min = round((time.time() - start_time) / 60, 1)
    print("\n--- TRADUCCION COMPLETADA ---")
    print(f"  Traducidos: {processed}")
    print(f"  Errores:    {errors}")
    print(f"  Tiempo:     {elapsed_min} minutos")

except KeyboardInterrupt:
    print("\nProceso interrumpido por el usuario.")
finally:
    conn.close()
